boards/boardsource-blok/examples/blok_usb_serial.rs

```diff
@@ -0,0 +1,160 @@
+//! # USB Serial Example for the Blok
+//!
+//! Creates a USB Serial device on a blok board, with the USB driver running in
+//! the USB interrupt and all writes being done in the main loop with the usage
+//! of critical section.
+//!
+//! This will create a USB Serial device and then write to it.
+//! A loop will write its current loop number 10 times after which
+//! the Blok will reset to usb boot mode.
+//! If "stop" is read by the serial device, the Blok will also
+//! reset to usb boot mode.
+//!
+//! See the `Cargo.toml` file for Copyright and license details.
+
+#![no_std]
+#![no_main]
+
+use boardsource_blok::{entry, hal};
+use boardsource_blok::{
+    hal::{
+        clocks::{init_clocks_and_plls, Clock},
+        pac,
+        pac::interrupt,
+        timer::Timer,
+        watchdog::Watchdog,
+        Sio,
+    },
+    Pins, XOSC_CRYSTAL_FREQ,
+};
+use core::fmt::Write;
+use heapless::String;
+use panic_halt as _;
+use usb_device::{
+    bus::UsbBusAllocator, device::UsbDevice, device::UsbDeviceBuilder, device::UsbVidPid,
+};
+use usbd_serial::SerialPort;
+
+// shared with the interrupt
+static mut USB_DEVICE: Option<UsbDevice<hal::usb::UsbBus>> = None;
+static mut USB_BUS: Option<UsbBusAllocator<hal::usb::UsbBus>> = None;
+static mut USB_SERIAL: Option<SerialPort<hal::usb::UsbBus>> = None;
```

Comment:
Is this good practice? The serial interrupt example of the pico does also use these unsafe blocks.

Reply:
Both `UsbDevice` and `SerialPort` reference `UsbBus` which is why I was not able to store them in multiple `Mutexe`s.
Is this is a readable and viable solution to avoid the unsafe blocks?

```rust
// imports

// shared with the interrupt
static STATIC_USB: Mutex<RefCell<Option<StaticUsb>>> = Mutex::new(RefCell::new(None));

struct StaticUsb<'a> {
    usb_bus: UsbBusAllocator<hal::usb::UsbBus>,
    usb_device: Option<UsbDevice<'a, hal::usb::UsbBus>>,
    usb_serial: Option<SerialPort<'a, hal::usb::UsbBus>>,
}

#[entry]
fn main() -> ! {

    // clocks, sio, etc...

    let usb_bus = UsbBusAllocator::new(hal::usb::UsbBus::new(
        pac.USBCTRL_REGS,
        pac.USBCTRL_DPRAM,
        clocks.usb_clock,
        true,
        &mut pac.RESETS,
    ));

    let _ = critical_section::with(|cs| {
        STATIC_USB.replace(cs, Some(StaticUsb { usb_bus, usb_serial: None, usb_device: None }));

        let mut static_usb = STATIC_USB.take(cs).unwrap();
        
        let usb_serial = SerialPort::new(&static_usb.usb_bus);

        let usb_device = UsbDeviceBuilder::new(&static_usb.usb_bus, UsbVidPid(0x1209, 0x0001))
                    .product("serial port")
                    .device_class(2) // from: https://www.usb.org/defined-class-codes
                    .build();

        static_usb.usb_serial = Some(usb_serial);
        static_usb.usb_device = Some(usb_device);
    });

    // main loop
}

/// Writes to the serial port.
///
/// We do this with interrupts disabled, to avoid a race hazard with the USB IRQ.
fn write_serial(byte_array: &[u8]) {
    let _ = critical_section::with(|cs| {
        STATIC_USB.take(cs).map(|usb| {
                usb.usb_serial.map(|mut serial| serial.write(byte_array))
            })
    });
}

// interrupt
```

Reply:
`take` removes the value from its container (see [`Mutex::take`](https://docs.rs/critical-section/latest/critical_section/struct.Mutex.html#method.take) pointing to [`RefCell::take`](https://doc.rust-lang.org/nightly/core/cell/struct.RefCell.html#method.take).
You should use [`Mutex::borrow_ref_mut`](https://docs.rs/critical-section/latest/critical_section/struct.Mutex.html#method.borrow_ref_mut) instead.
But yes, that's the idea.

Reply:
I'll be creating a PR this evening (BST) to update the rp-pico's usb_serial_interrupt example to show a more comprehensive and up-to-date example.

Reply:
See: https://github.com/rp-rs/rp-hal-boards/pull/37

Reply:
Thank you very much